File: android-app/app/src/main/java/com/petdocs/android/data/SessionStore.kt
package com.petdocs.android.data

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

private val Context.petdocsStore by preferencesDataStore("petdocs_session")

/**
 * Lightweight session for the Android app. Mirrors the portal's SessionStore
 * shape but with `petdocs_*` keys: the owner's email (for display + upload
 * attribution), the Convex `owners` id (required arg for every query), and an
 * optional Convex URL override (defaults to BuildConfig.CONVEX_URL).
 */
class SessionStore(private val context: Context) {
    private val ownerEmailKey = stringPreferencesKey("petdocs_owner_email")
    private val ownerIdKey = stringPreferencesKey("petdocs_owner_id")
    private val convexUrlOverrideKey = stringPreferencesKey("petdocs_convex_url")

    val ownerEmail: Flow<String?> = context.petdocsStore.data.map { prefs ->
        prefs[ownerEmailKey]?.ifBlank { null }
    }

    val ownerId: Flow<String?> = context.petdocsStore.data.map { prefs ->
        prefs[ownerIdKey]?.ifBlank { null }
    }

    val convexUrlOverride: Flow<String?> = context.petdocsStore.data.map { prefs ->
        prefs[convexUrlOverrideKey]?.ifBlank { null }
    }

    suspend fun setOwnerEmail(email: String?) {
        context.petdocsStore.edit { prefs ->
            if (email.isNullOrBlank()) prefs.remove(ownerEmailKey) else prefs[ownerEmailKey] = email
        }
    }

    suspend fun setOwnerId(ownerId: String?) {
        context.petdocsStore.edit { prefs ->
            if (ownerId.isNullOrBlank()) prefs.remove(ownerIdKey) else prefs[ownerIdKey] = ownerId
        }
    }

    suspend fun setConvexUrlOverride(url: String?) {
        context.petdocsStore.edit { prefs ->
            if (url.isNullOrBlank()) prefs.remove(convexUrlOverrideKey) else prefs[convexUrlOverrideKey] = url.trim()
        }
    }

    suspend fun setSession(email: String?, ownerId: String?) {
        context.petdocsStore.edit { prefs ->
            if (email.isNullOrBlank()) prefs.remove(ownerEmailKey) else prefs[ownerEmailKey] = email
            if (ownerId.isNullOrBlank()) prefs.remove(ownerIdKey) else prefs[ownerIdKey] = ownerId
        }
    }

    suspend fun clear() {
        context.petdocsStore.edit { it.clear() }
    }
}

File: ios-app/PetDocs/SessionStore.swift
import Foundation

// Lightweight session for the iOS app. Mirrors the Android SessionStore
// shape: the owner email (display plus upload attribution), the Convex
// owners id (required arg for every query), and an optional Convex URL
// override. Stored in the "petdocs" UserDefaults suite.
final class SessionStore: ObservableObject {
    @Published var ownerEmail: String?
    @Published var ownerId: String?
    @Published var convexUrl: String?

    private let defaults: UserDefaults
    private enum Keys {
        static let email = "petdocs_owner_email"
        static let ownerId = "petdocs_owner_id"
        static let convexUrl = "petdocs_convex_url"
    }

    init(defaults: UserDefaults = UserDefaults(suiteName: "petdocs") ?? .standard) {
        self.defaults = defaults
        ownerEmail = defaults.string(forKey: Keys.email)
        ownerId = defaults.string(forKey: Keys.ownerId)
        convexUrl = defaults.string(forKey: Keys.convexUrl)
    }

    var isSignedIn: Bool {
        !(ownerId ?? "").isEmpty
    }

    // Backend URL: per-device override wins, otherwise the build default.
    func resolvedConvexUrl(fallback: String) -> String {
        let override = (convexUrl ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return override.isEmpty ? fallback : override
    }

    func save(email: String?, ownerId: String?) {
        set(Keys.email, email)
        set(Keys.ownerId, ownerId)
        ownerEmail = email
        self.ownerId = ownerId
    }

    func setConvexUrl(_ url: String?) {
        let trimmed = url?.trimmingCharacters(in: .whitespacesAndNewlines)
        set(Keys.convexUrl, (trimmed ?? "").isEmpty ? nil : trimmed)
        convexUrl = (trimmed ?? "").isEmpty ? nil : trimmed
    }

    func clear() {
        defaults.removeObject(forKey: Keys.email)
        defaults.removeObject(forKey: Keys.ownerId)
        defaults.removeObject(forKey: Keys.convexUrl)
        ownerEmail = nil
        ownerId = nil
        convexUrl = nil
    }

    private func set(_ key: String, _ value: String?) {
        if let value, !value.isEmpty {
            defaults.set(value, forKey: key)
        } else {
            defaults.removeObject(forKey: key)
        }
    }
}
